builtin/credential/aws/path_role.go

```diff
@@ -320,6 +322,64 @@ func (b *backend) setRole(ctx context.Context, s logical.Storage, roleName strin
 	return nil
 }
 
+// updateUpgradableRoleEntries upgrades and persists all of the role entries
+// that are in need of being upgraded.
+func (b *backend) updateUpgradableRoleEntries(ctx context.Context, req *logical.InitializationRequest) error {
+
+	// Upgrade only if we are either: (1) a local mount, or (2) are _not_ a
+	// performance replicated standby cluster.
+	if !(b.System().LocalMount() || !b.System().ReplicationState().HasState(
+		consts.ReplicationPerformanceSecondary|consts.ReplicationPerformanceStandby)) {
+		return nil
+	}
+
+	// Read all the role names.
+	b.roleMutex.RLock()
+	roleNames, err := req.Storage.List(ctx, "role/")
+	b.roleMutex.RUnlock()
+	if err != nil {
+		return err
+	}
+
+	// Upgrade the roles as necessary.
+	for _, roleName := range roleNames {
+		err := b.updateUpgradableRoleEntry(ctx, req.Storage, roleName)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// updateUpgradableRoleEntry uses the write lock to read a role and persist it
+// if it needs to be upgraded.
+func (b *backend) updateUpgradableRoleEntry(ctx context.Context, s logical.Storage, roleName string) error {
```

Comment:
I don't think this function is actually needed at all; I think it could be replaced in the above function with grabbing the write lock and then having the range loop just all roleInternal.

Reply:
this function is gone now

Reply:
actually wait, its still there because I grab the write lock per-role and just call `roleInternal()`.  I think you are saying that I can just hold the lock before the range loop?  I believe you suggested at first not to do that because it could be a long process to upgrade all the roles and we didn't want to hold the lock open that long.  I'd be happy to change it though